File: java/src/main/java/org/urlshortener/dal/repositories/ShortUrlRepository.java
package org.urlshortener.dal.repositories;

import lombok.NonNull;
import org.urlshortener.dal.entities.ShortUrl;

import javax.validation.Valid;
import java.util.List;

public interface ShortUrlRepository {
    ShortUrl merge(@NonNull @Valid ShortUrl shortUrl);

    void deleteById(@NonNull Long id);

    Long countByShortName(@NonNull String shortName);

    ShortUrl findByShortName(@NonNull String shortName);

    ShortUrl findByIdAndUserId(Long id, String userId);

    boolean existsByIdAndUserId(Long id, String userId);

    List<ShortUrl> findAllByUserId(String userId);
}

File: kotlin/src/main/kotlin/com/urlshortener/dal/repositories/ShortUrlRepository.kt
package com.urlshortener.dal.repositories

import com.urlshortener.dal.entities.ShortUrl
import javax.validation.Valid

interface ShortUrlRepository {
    fun merge(@Valid shortUrl: ShortUrl): ShortUrl
    fun deleteById(id: Long)
    fun countByShortName(shortName: String): Long
    fun findByShortName(shortName: String): ShortUrl?
    fun findByIdAndUserId(id: Long, userId: String): ShortUrl?
    fun existsByIdAndUserId(id: Long, userId: String): Boolean
    fun findAllByUserId(userId: String): List<ShortUrl>
}